File: baml_language/sdk_tests/crates/cpp/type_shapes/customizable/tests/test_media.cc
// Roundtrip coverage for the media suite.
// Port of type_shapes/customizable/roundtrip_tests/test_media.py. Media
// values can't be hand-built as plain structs with meaningful content, so
// each value is sourced from the matching return_* function (which builds
// it engine-side via image.from_url(...) etc.); the decode path yields a
// baml.media.* class wrapping a _data handle, and the encode path passes
// that value back into a round_trip_* function.
#include <baml_sdk.h>
#include <baml_test.h>

#include <optional>
#include <string>

namespace media_ns = baml_sdk::media;
namespace baml_media = baml_sdk::baml::media;

static const char kUrl[] = "https://example.com/asset";

// --- decode path (return_*) works -----------------------------------------

BAML_TEST(return_image) {
  const baml_media::Image img = media_ns::return_image(kUrl, std::nullopt);
  BAML_ASSERT(!img._data.empty());
  BAML_ASSERT(img.url() == std::optional<std::string>(kUrl));
}

BAML_TEST(return_audio) {
  const baml_media::Audio aud = media_ns::return_audio(kUrl, std::nullopt);
  BAML_ASSERT(!aud._data.empty());
}

BAML_TEST(return_video) {
  const baml_media::Video vid = media_ns::return_video(kUrl, std::nullopt);
  BAML_ASSERT(!vid._data.empty());
}

BAML_TEST(return_pdf) {
  const baml_media::Pdf pdf = media_ns::return_pdf(kUrl, std::nullopt);
  BAML_ASSERT(!pdf._data.empty());
}

// --- encode path (round_trip_*) --------------------------------------------

BAML_TEST(round_trip_image) {
  const baml_media::Image img = media_ns::return_image(kUrl, std::nullopt);
  const baml_media::Image back = media_ns::round_trip_image(img);
  BAML_ASSERT(!back._data.empty());
  BAML_ASSERT(back.url() == std::optional<std::string>(kUrl));
}

BAML_TEST(round_trip_audio) {
  const baml_media::Audio aud = media_ns::return_audio(kUrl, std::nullopt);
  BAML_ASSERT(!media_ns::round_trip_audio(aud)._data.empty());
}

BAML_TEST(round_trip_video) {
  const baml_media::Video vid = media_ns::return_video(kUrl, std::nullopt);
  BAML_ASSERT(!media_ns::round_trip_video(vid)._data.empty());
}

BAML_TEST(round_trip_pdf) {
  const baml_media::Pdf pdf = media_ns::return_pdf(kUrl, std::nullopt);
  BAML_ASSERT(!media_ns::round_trip_pdf(pdf)._data.empty());
}

BAML_TEST(round_trip_media) {
  const media_ns::Media m{
      media_ns::return_image(kUrl, std::nullopt),
      media_ns::return_audio(kUrl, std::nullopt),
      media_ns::return_video(kUrl, std::nullopt),
      media_ns::return_pdf(kUrl, std::nullopt),
  };
  const media_ns::Media back = media_ns::round_trip_media(m);
  BAML_ASSERT(!back.image_field._data.empty());
  BAML_ASSERT(!back.audio_field._data.empty());
  BAML_ASSERT(!back.video_field._data.empty());
  BAML_ASSERT(!back.pdf_field._data.empty());
}

File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/roundtrip_tests/test_media.rs
//! Roundtrip coverage for `baml_sdk::media`.
//!
//! Values exercise both outbound descriptors returned by BAML and Rust-side
//! media constructors encoded through the bridge's native handle ABI.
use baml_bridge::{
    DecodeError, Error,
    media::{Audio, Image},
};
use baml_sdk::media::{
    ImageOrAudio, Media, return_audio, return_image, return_pdf, return_video, round_trip_audio,
    round_trip_image, round_trip_image_or_audio, round_trip_media, round_trip_pdf,
    round_trip_video,
};

const URL: &str = "https://example.com/asset";

// --- decode path (return_*) works -----------------------------------------

#[test]
fn test_media_return_image() {
    // DIVERGENCE(rust): python asserts `is not None`; the successful unwrap
    // of the non-optional media result is that assertion here (and in every
    // test below).
    return_image(URL.to_string(), None).unwrap();
}

#[test]
fn test_media_return_audio() {
    return_audio(URL.to_string(), None).unwrap();
}

#[test]
fn test_media_return_video() {
    return_video(URL.to_string(), None).unwrap();
}

#[test]
fn test_media_return_pdf() {
    return_pdf(URL.to_string(), None).unwrap();
}

// --- encode path (round_trip_*) ------------------------------------------

#[test]
fn test_media_round_trip_image() {
    let path = std::env::temp_dir().join("baml-rust-sdk-media-example.png");
    let path = path.to_string_lossy().into_owned();
    let img = Image::from_file(path.clone(), Some("image/png".to_string())).unwrap();
    let returned = round_trip_image(img).unwrap();
    assert_eq!(returned.file(), Some(path.as_str()));
    assert_eq!(returned.mime_type(), Some("image/png"));
}

#[test]
fn test_media_round_trip_audio() {
    let aud = return_audio(URL.to_string(), None).unwrap();
    round_trip_audio(aud).unwrap();
}

#[test]
fn test_media_round_trip_video() {
    let vid = return_video(URL.to_string(), None).unwrap();
    round_trip_video(vid).unwrap();
}

#[test]
fn test_media_round_trip_pdf() {
    let pdf = return_pdf(URL.to_string(), None).unwrap();
    round_trip_pdf(pdf).unwrap();
}

#[test]
fn test_media_round_trip_media() {
    let m = Media {
        image_field: return_image(URL.to_string(), None).unwrap(),
        audio_field: return_audio(URL.to_string(), None).unwrap(),
        video_field: return_video(URL.to_string(), None).unwrap(),
        pdf_field: return_pdf(URL.to_string(), None).unwrap(),
    };
    round_trip_media(m).unwrap();
}

// SDK_PARITY_LINT(skip): validates Rust-specific generated media-union encode/decode
#[test]
fn test_media_round_trip_union() {
    let img = Image::from_url(URL, Some("image/png".to_string())).unwrap();
    let returned = round_trip_image_or_audio(ImageOrAudio::Image(img)).unwrap();
    assert!(matches!(returned, ImageOrAudio::Image(_)));

    let audio = Audio::from_url(URL, Some("audio/mpeg".to_string())).unwrap();
    let returned = round_trip_image_or_audio(ImageOrAudio::Audio(audio)).unwrap();
    assert!(matches!(returned, ImageOrAudio::Audio(_)));
}

// SDK_PARITY_LINT(skip): validates Rust bridge rejection of C-incompatible media descriptors
#[test]
fn test_media_return_rejects_interior_nul() {
    let error = return_image("bad\0url".to_string(), None).unwrap_err();
    assert!(matches!(
        error,
        Error::Decode(DecodeError::InvalidMedia { field: "source" })
    ));
}
